### PR TITLE
Use merging fieldsreader when restoring versionmap during recovery

### PR DESCRIPTION
**Pull request by romseygeek**
When reading sequential data from compressed stored fields it is 
more efficient to use a CodecReader's merge instance, as this holds
decompressed blocks in memory and avoids re-doing the 
decompression for every document.  When restoring the version map
during a recovery, we read id fields from all documents beyond the global
checkpoint, and this is precisely the situation where this performance
enhancement should help.

This commit updates the IdOnlyFieldsVisitor to use a merge instance, 
and in the process tidies up the API a bit; we now have a package-private 
IdStoredFieldLoader that handles fetching merge instances and loading 
ids for each document.

### CODE REVIEW

**Review by dnhatn — APPROVED** [review]
LGTM. The change makes sense to me. Thanks, Alan.

File: server/src/main/java/org/elasticsearch/index/engine/IdStoredFieldLoader.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+    /**
+     * Creates a new ID reader over a segment
+     */
+    IdStoredFieldLoader(LeafReader in) {
+        this.reader = getStoredFieldsReader(in);
+    }
+
+    private static CheckedBiConsumer<Integer, StoredFieldVisitor, IOException> getStoredFieldsReader(LeafReader in) {
+        if (in instanceof CodecReader) {
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/index/engine/IdStoredFieldLoader.java:48]: We wrap the directory reader in # so this is never true.  We don't use codec readers, instead we introduced a `SequentialStoredFieldsLeafReader` that exposes the merge instance lightly (see [`SourceLookup`]()).
That avoids rewriting all filter leaf readers as codec readers. And so in this case the [LeafReaderWithLiveDocs] should extend `SequentialStoredFieldsLeafReader` to expose this functionality.

**Review by jimczi — CHANGES_REQUESTED** [review]
I left one comment, the logic to access the merge instance seems wrong.

**Comment by romseygeek** [comment]
@elasticmachine update branch

**Comment by romseygeek** [comment]
@elasticmachine run elasticsearch-ci/1

File: server/src/main/java/org/elasticsearch/common/lucene/index/SequentialStoredFieldsLeafReader.java [hunk 2]
```diff
@@ -53,15 +51,15 @@ public SequentialStoredFieldsLeafReader(LeafReader in) {
...
+    public StoredFieldsReader getSequentialStoredFieldsReader() {
         if (in instanceof CodecReader) {
             CodecReader reader = (CodecReader) in;
             return doGetSequentialStoredFieldsReader(reader.getFieldsReader().getMergeInstance());
         } else if (in instanceof SequentialStoredFieldsLeafReader) {
             SequentialStoredFieldsLeafReader reader = (SequentialStoredFieldsLeafReader) in;
             return doGetSequentialStoredFieldsReader(reader.getSequentialStoredFieldsReader());
         } else {
-            throw new IOException("requires a CodecReader or a SequentialStoredFieldsLeafReader, got " + in.getClass());
+            throw new IllegalArgumentException("requires a CodecReader or a SequentialStoredFieldsLeafReader, got " + in.getClass());
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/common/lucene/index/SequentialStoredFieldsLeafReader.java:62]: An `IllegalStateException` seems more appropriate ?

**Review by jimczi — APPROVED** [review]
LGTM, thanks

File: server/src/main/java/org/elasticsearch/index/engine/IdStoredFieldLoader.java (hunk 1 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/engine/IdStoredFieldLoader.java]: Good catch! Thanks Jim!

**Comment by romseygeek** [comment]
@elasticmachine update branch

**Pull request merged**